chronos/futures.nim

```diff
@@ -0,0 +1,369 @@
+    ## may not be familiar with, in particular calls inside the stdlib.
+
+    let (filename, procname) = getFilenameProcname(entry)
+
+    if procname == "processPendingCallbacks":
+      if cmpIgnoreStyle(filename, "asyncdispatch.nim") == 0:
+        return "Executes pending callbacks"
+    elif procname == "poll":
+      if cmpIgnoreStyle(filename, "asyncdispatch.nim") == 0:
+        return "Processes asynchronous completion events"
+
+    if procname == "internalContinue":
+      if cmpIgnoreStyle(filename, "asyncfutures.nim") == 0:
+        return "Resumes an async procedure"
+
+  proc `$`(stackTraceEntries: seq[StackTraceEntry]): string =
+    try:
+      when defined(nimStackTraceOverride) and declared(addDebuggingInfo):
+        let entries = addDebuggingInfo(stackTraceEntries)
+      else:
+        let entries = stackTraceEntries
+
+      # Find longest filename & line number combo for alignment purposes.
+      var longestLeft = 0
+      for entry in entries:
+        let (filename, procname) = getFilenameProcname(entry)
+
+        if procname == "": continue
+
+        let leftLen = filename.len + len($entry.line)
+        if leftLen > longestLeft:
+          longestLeft = leftLen
+
+      var indent = 2
+      # Format the entries.
+      for entry in entries:
+        let (filename, procname) = getFilenameProcname(entry)
+
+        if procname == "":
+          if entry.line == reraisedFromBegin:
+            result.add(spaces(indent) & "#[\n")
+            indent.inc(2)
+          elif entry.line == reraisedFromEnd:
+            indent.dec(2)
+            result.add(spaces(indent) & "]#\n")
+          continue
+
+        let left = "$#($#)" % [filename, $entry.line]
+        result.add((spaces(indent) & "$#$# $#\n") % [
+          left,
+          spaces(longestLeft - left.len + 2),
+          procname
+        ])
+        let hint = getHint(entry)
+        if hint.len > 0:
+          result.add(spaces(indent+2) & "## " & hint & "\n")
+    except ValueError as exc:
+      return exc.msg # Shouldn't actually happen since we set the formatting
+                    # string
+
+  proc injectStacktrace(future: FutureBase) =
+    const header = "\nAsync traceback:\n"
+
+    var exceptionMsg = future.error.msg
+    if header in exceptionMsg:
+      # This is messy: extract the original exception message from the msg
+      # containing the async traceback.
+      let start = exceptionMsg.find(header)
+      exceptionMsg = exceptionMsg[0..<start]
+
+    var newMsg = exceptionMsg & header
+
+    let entries = getStackTraceEntries(future.error)
+    newMsg.add($entries)
+
+    newMsg.add("Exception message: " & exceptionMsg & "\n")
+
+    # # For debugging purposes
+    # newMsg.add("Exception type:")
+    # for entry in getStackTraceEntries(future.error):
+    #   newMsg.add "\n" & $entry
+    future.error.msg = newMsg
+
+# Internal utilities - these are not part of the stable API
+proc internalCheckComplete*(fut: FutureBase) {.
```

Comment:
This thing should not be `procedure` it should be `template` otherwise it will always fall into stack traces and sometimes it hides original procedure name, so even for further work on async traces this should become `template`.

Reply:
it might be better to keep this as a function - it doesn't make any difference performance-wise really (compared to all allocations going on in general), but turning it into a function means we'll have a copy of the code in every instantiation (instead of a single one for `FutureBase`) - chronos is already one of the biggest code size contributors, so I think it might make sense to keep this as a non-generic proc

Reply:
Both `proc` and `func` do not solve the problem of forming stack/async traces. And getting proper stack/async trace is much more important... Right now we don't have ability to form this stack/async trace, but when this work will be started all this "helper" procedures/funcs will hide place in actual async procedure code. So this "helpers" should become hidden and "helpers" become hidden when they become templates.

Reply:
made it a template for now - I think I found a better fix for stack code locations but I want to merge this PR first